tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs

```diff
@@ -50,18 +54,70 @@ private SpanMessagePackFormatter()
             _processIdValueBytes = processId > 0 ? MessagePackSerializer.Serialize(processId) : null;
         }
 
+        // IMessagePackFormatter<ArraySegment<Span>> is deprecated, use IMessagePackFormatter<TraceMessageChunk>
+        public int Serialize(ref byte[] bytes, int offset, ArraySegment<Span> value, IFormatterResolver formatterResolver)
+        {
+            var traceContext = value.Array![value.Offset].Context.TraceContext;
+            var traceChunk = new TraceChunkModel(value, traceContext);
+            return Serialize(ref bytes, offset, traceChunk, formatterResolver);
+        }
+
+        public int Serialize(ref byte[] bytes, int offset, TraceChunkModel value, IFormatterResolver formatterResolver)
+        {
+            int originalOffset = offset;
+            var spans = value.Spans;
+            var spanIds = new HashSet<ulong>(spans.Count);
```

Comment:
tiny optimisation, but I wonder, for the small span counts we see generally does it make sense to have a list? Also removes the (miniscule) chance of collisions narrowing `ulong` I guess. But worse for big traces so 🤷‍♂️

Reply:
Agreed! I'll look into this.

Reply:
I added a new `SpanIdLookup` type that switches between either iterating the `ArraySegment<Span>` directly when there are < 50 spans or building a `HashSet<ulong>` otherwise.

I also check the first span first before doing anything else which short-circuits the common cases:
- single-span trace chunks
- root first span is usually the local root, and most spans are children of the root